### PR TITLE
feat(data-request): include collateral in tally

### LINKED ISSUE #1113: [Collaterals]: Tally Transactions

### PR DESCRIPTION
**Pull request by lrubiorod**
Close #1113

### CODE REVIEW

File: data_structures/src/data_request.rs [hunk 1]
```diff
@@ -357,6 +356,27 @@ pub fn calculate_tally_change(
...
+pub fn calculate_reward(
+    reveals_count: usize,
+    slashed_count: usize,
+    honests_count: usize,
+    dr_output: &DataRequestOutput,
+) -> (u64, u64) {
+    if reveals_count > 0 {
+        let slashed_count = slashed_count as u64;
+        let honests_count = honests_count as u64;
+        let slashed_collateral_reward = dr_output.collateral * slashed_count / honests_count;
```
Comment (tmpolaczyk) [inline data_structures/src/data_request.rs]: `honests_count` can be 0?

File: data_structures/src/data_request.rs [hunk 2]
```diff
@@ -384,31 +403,52 @@ where
...
-            debug!("Created tally change for the data request creator");
-            let tally_change =
-                calculate_tally_change(commits_count, reveals_count, honests_count, &dr_output);
+        let liars_count = liars.iter().fold(0, |count, liar| match liar {
+            true => count + 1,
+            false => count,
+        });
+        let slashed_count = commits_count - reveals_count + liars_count;
+        let honests_count = reveals_count - liars_count;
+        // Collateral division rest goes for the miner
```
Comment (tmpolaczyk) [inline data_structures/src/data_request.rs]: This is not implemented yet

File: data_structures/src/data_request.rs (hunk 1 above)
Comment (lrubiorod) [reply data_structures/src/data_request.rs]: In case of reveals_count > 0 it is not possible

File: data_structures/src/data_request.rs (hunk 2 above)
Comment (lrubiorod) [reply data_structures/src/data_request.rs]: It is, in validate_tally_transaction, collateral rest is included in the tally fee

File: data_structures/src/data_request.rs (hunk 1 above)
Comment (tmpolaczyk) [reply data_structures/src/data_request.rs]: What if everyone lies?

File: data_structures/src/data_request.rs [hunk 3]
```diff
@@ -357,6 +356,30 @@ pub fn calculate_tally_change(
...
+    if commits_count == 0 {
+        (0, 0)
+    } else if reveals_count == 0 {
+        (dr_output.collateral, 0)
+    } else {
+        let liars_count = reveals_count - honests_count;
+        let slashed_count = (commits_count - reveals_count + liars_count) as u64;
+        let honests_count = honests_count as u64;
+        let slashed_collateral_reward = dr_output.collateral * slashed_count / honests_count;
+        let slashed_collateral_rest = (dr_output.collateral * slashed_count) % honests_count;
```
Comment (aesedepece) [inline data_structures/src/data_request.rs]:
```suggestion
        let slashed_collateral_remainder = (dr_output.collateral * slashed_count) % honests_count;
```

File: validations/src/validations.rs [hunk 4]
```diff
@@ -986,20 +994,20 @@ pub fn validate_tally_transaction<'a>(
...
         } else {
             // Validation of each rewarded address was a revealer
-            if dr_state.info.reveals.get(&output.pkh).is_none() {
+            if reveals_count > 0 && dr_state.info.reveals.get(&output.pkh).is_none() {
                 return Err(TransactionError::RevealNotFound.into());
             }
 
             // Validation of each rewarded address was not a liar
-            if slashed_hs.contains(&output.pkh) {
+            if reveals_count > 0 && slashed_hs.contains(&output.pkh) {
```
Comment (aesedepece) [inline validations/src/validations.rs]: The expression evaluated first in both conditionals is the same. I'd favor nested conditionals:

```rust
if reveals_count > 0 {
    // Make sure every rewarded address is a revealer
    if dr_state.info.reveals.get(&output.pkh).is_none() {
        return Err(TransactionError::RevealNotFound.into());
    }
    // Make sure every rewarded address passed the tally function, a.k.a. "is honest" / "is not a liar"
    if slashed_hs.contains(&output.pkh) {
        return Err(TransactionError::DishonestReward.into());
    }
}
```

File: data_structures/src/data_request.rs (hunk 2 above)
Comment (tmpolaczyk) [reply data_structures/src/data_request.rs]: Oh, I looked at mining.rs and the fee is not counted here

File: data_structures/src/data_request.rs (hunk 1 above)
Comment (lrubiorod) [reply data_structures/src/data_request.rs]: Our honests people is people in consensus. If all lies are in consensus, so all of them are honest people for us

File: node/src/actors/chain_manager/mining.rs [hunk 5]
```diff
@@ -705,9 +705,19 @@ fn build_block(
     }
 
     for ta_tx in tally_transactions {
-        if let Some(dr_output) = dr_pool.get_dr_output(&ta_tx.dr_pointer) {
+        if let Some(dr_state) = dr_pool.data_request_state(&ta_tx.dr_pointer) {
             tally_txns.push(ta_tx.clone());
-            transaction_fees += dr_output.tally_fee;
+            let commits_count = dr_state.info.commits.len();
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/mining.rs]: This repeated logic makes me think that `calculate_witness_reward` should be a method of `dr_state`, so we can just do `dr_state.calculate_witness_reward(&dr_pointer)`

Comment (lrubiorod) [reply node/src/actors/chain_manager/mining.rs:710]: honests_count can not be calculated with dr_state information

Comment (tmpolaczyk) [reply node/src/actors/chain_manager/mining.rs:710]: Then what about `dr_state.calculate_witness_reward(&dr_pointer, &tally_transaction)`

Comment (lrubiorod) [reply node/src/actors/chain_manager/mining.rs:710]: There are another functions where is used (i.e. create_tally), where do not have access to dr_state

File: validations/src/validations.rs [hunk 6]
```diff
@@ -12,6 +12,7 @@ use witnet_crypto::{
     merkle::{merkle_tree_root as crypto_merkle_tree_root, ProgressiveMerkleTree},
     signature::{verify, PublicKey, Signature},
 };
+use witnet_data_structures::data_request::calculate_witness_reward;
```
Comment (tmpolaczyk) [inline validations/src/validations.rs:15]: fmt

File: data_structures/src/data_request.rs [hunk 7]
```diff
@@ -357,6 +356,30 @@ pub fn calculate_tally_change(
...
+    if commits_count == 0 {
+        (0, 0)
+    } else if reveals_count == 0 {
+        (dr_output.collateral, 0)
+    } else {
+        let liars_count = reveals_count - honests_count;
+        let slashed_count = (commits_count - reveals_count + liars_count) as u64;
+        let honests_count = honests_count as u64;
+        let slashed_collateral_reward = dr_output.collateral * slashed_count / honests_count;
+        let slashed_collateral_reminder = (dr_output.collateral * slashed_count) % honests_count;
```
Comment (tmpolaczyk) [inline data_structures/src/data_request.rs:374]:
```suggestion
        let slashed_collateral_remainer = (dr_output.collateral * slashed_count) % honests_count;
```

:laughing:

Comment (aesedepece) [reply data_structures/src/data_request.rs:374]: It's not `reminder` nor `remainer` but `remainder`:
https://en.wikipedia.org/wiki/Remainder

**Pull request merged**